programs/ziti-edge-tunnel/netif_driver/linux/resolvers.c

```diff
@@ -0,0 +1,408 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#ifdef USE_LIBSYSTEMD_RESOLVER
+#include <systemd/sd-bus-protocol.h>
+#include <systemd/sd-daemon.h>
+#include <systemd/sd-bus.h>
+#endif
+
+#include <ziti/ziti_log.h>
+#include <uv.h>
+
+#include "utils.h"
+
+#ifdef USE_LIBSYSTEMD_RESOLVER
+#ifndef RESOLVED_DBUS_NAME
+#define RESOLVED_DBUS_NAME "org.freedesktop.resolve1"
+#endif
+
+#ifndef RESOLVED_DBUS_PATH
+#define RESOLVED_DBUS_PATH "/org/freedesktop/resolve1"
+#endif
+
+#ifndef RESOLVED_DBUS_MANAGER_INTERFACE
+#define RESOLVED_DBUS_MANAGER_INTERFACE "org.freedesktop.resolve1.Manager"
+#endif
+
+#define _cleanup_(f) __attribute__((cleanup(f)))
+#define TRY_DL(dl_func) do{if ((dl_func) != 0) goto dl_error;} while(0)
+
+static int detect_systemd_resolved_acquired_name(sd_bus *bus);
+static int detect_systemd_resolved_routing_domain_wildcard(sd_bus *bus, int32_t ifindex);
+static bool set_systemd_resolved_link_dns(sd_bus *bus, int32_t ifindex, const char* tun, const char* dns_addr);
+static bool set_systemd_resolved_link_dnssec(sd_bus *bus, int32_t ifindex, const char*tun);
+static bool set_systemd_resolved_link_domain(sd_bus *bus, int32_t ifindex, const char* tun);
+
+// libsystemd prototypes
+static uv_once_t guard;
+static uv_lib_t libsystemd_h;
+static int (*sd_booted_f)(void);
+static int (*sd_bus_call_f)(sd_bus *bus, sd_bus_message *m, uint64_t usec, sd_bus_error *ret_error, sd_bus_message **reply);
+static int (*sd_bus_call_method_f)(sd_bus *bus, const char *destination, const char *path, const char *interface, const char *member, sd_bus_error *ret_error, sd_bus_message **reply, const char *types, ...);
+static void (*sd_bus_error_free_f)(sd_bus_error *e);
+static sd_bus *(*sd_bus_flush_close_unref_f)(sd_bus *bus);
+static int (*sd_bus_get_property_f)(sd_bus *bus, const char *destination, const char *path, const char *interface, const char *member, sd_bus_error *ret_error, sd_bus_message **reply, const char *type);
+static int (*sd_bus_is_bus_client_f)(sd_bus *bus);
+static int (*sd_bus_message_append_f)(sd_bus_message *m, const char *types, ...);
+static int (*sd_bus_message_append_array_f)(sd_bus_message *m, char type, void *ptr, size_t size);
+static int (*sd_bus_message_close_container_f)(sd_bus_message *m);
+static int (*sd_bus_message_enter_container_f)(sd_bus_message *m, char type, const char *contents);
+static int (*sd_bus_message_exit_container_f)(sd_bus_message *m);
+static int (*sd_bus_message_new_method_call_f)(sd_bus *bus, sd_bus_message **m, const char *destination, const char *path, const char *interface, const char *member);
+static int (*sd_bus_message_open_container_f)(sd_bus_message *m, char type, const char *contents);
+static int (*sd_bus_message_read_f)(sd_bus_message *m, const char *types, ...);
+static int (*sd_bus_message_read_strv_f)(sd_bus_message *m, char ***l);
+static sd_bus_message *(*sd_bus_message_unref_f)(sd_bus_message *m);
+static int (*sd_bus_open_system_f)(sd_bus **bus);
+
+static void init_libsystemd() {
```

Comment:
Why are you mapping symbols one-at-a-time with dlsym? Is there any reason not to add -lsystemd to the link libraries and reference the symbols directly?

Reply:
This aims to make the shared library an optional runtime dependency. `-lsystemd` is already added to the target link libraries at build time, provided a valid (found and version >=229) library is found. Am I misunderstanding the implications of these actions here?

Reply:
> Why are you mapping symbols one-at-a-time with dlsym? Is there any reason not to add -lsystemd to the link libraries and reference the symbols directly?

we cannot require linking against systemd

Reply:
There is certainly room for a misunderstanding on my part :)

I didn't think about making the library an optional runtime dependency. It makes sense if a system may not have libsystemd.so on the library path. Using `dlopen` and `dlsym` does achieve that on its own, however I believe that the runtime dependency becomes required when you _also_ link with `-lsystemd`.

Have you tried not explicitly linking libsystemd?

Reply:
You are absolutely correct. While the initial intention was there, we were indeed (accidentally) linking the library. This was a development process artifact. This should be addressed now with the latest commit.